xarray/core/dataset.py

```diff
@@ -6294,5 +6294,55 @@ def idxmax(
             ),
         )
 
+    def argmin(self, dim=None, axis=None, **kwargs):
+        if dim is None and axis is None:
+            warnings.warn(
+                "Behaviour of DataArray.argmin() with neither dim nor axis argument "
+                "will change to return a dict of indices of each dimension, and then it "
+                "will be an error to call Dataset.argmin() with no argument. To get a "
+                "single, flat index, please use np.argmin(ds) instead of ds.argmin().",
+                DeprecationWarning,
```

Comment:
When issuing warnings, please add the `stacklevel` keyword argument, which ensures that the warning ends up in user code where they call the xarray method, rather than inside the method. Usually `stacklevel=2` is right:
https://docs.python.org/3/library/warnings.html#warnings.warn

Sometimes you need a higher `stacklevel` if the warning is issued from a helper function